api/interceptor/observer.go

```diff
@@ -0,0 +1,269 @@
+			peerAddress: r.RemoteAddr,
+			protocol:    "http",
+			procedure:   procedure,
+			requestURI:  r.RequestURI,
+			method:      r.Method,
+			// Other properties are set by wrapped body ready, response writer
+			// as well as connectrpc interceptors.
+		}
+		ctx := setRequestMetadata(r.Context(), rm)
+		r = r.WithContext(ctx)
+
+		// 2. Inject custom body reader that counts the bytes as part of the request
+		r.Body = bodyReader{
+			base:      r.Body,
+			bytesRead: &rm.bytesReceived,
+		}
+
+		// 3. Wrap a response writer to collect metrics while writing the response
+		wrappedResponseWriter := responseController{
+			base: w.(responseWriteFlusher),
+			rm:   rm,
+		}
+
+		// 4. Let handler(s) process the request
+		next.ServeHTTP(wrappedResponseWriter, r)
+
+		// 5. Inform RequestMetadata that we can stop measuring duration
+		rm.end()
+		in.onRequestEnded(r.Context(), rm)
+	})
+}
+
+// WrapUnary creates an interceptor to validate Connect requests.
+func (*Observer) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		if req.Spec().IsClient {
+			return next(ctx, req)
+		}
+
+		rm := getRequestMetadata(ctx)
+
+		// For HTTP paths invoked via gRPC gateway, procedure is expected not to be set.
+		// In that case, let's try to retrieve it with gRPC gateway's runtime pkg.
+		procedure := req.Spec().Procedure
+		if procedure == "" {
+			path, ok := runtime.RPCMethod(ctx)
+			if !ok {
+				procedure = "unknown"
+			} else {
+				procedure = path
+			}
+		}
+		rm.procedure = procedure
+
+		protocol := req.Peer().Protocol
+		if protocol == "" {
+			protocol = "http"
+		}
+		rm.protocol = protocol
+
+		// 2. Execute request
+		response, err := next(ctx, req)
+		rm.err = err
+
+		return response, err
+	}
+}
+
+// WrapStreamingClient is the middleware handler for bidirectional requests from
+// the client perspective.
+func (*Observer) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
+	return next
+}
+
+// WrapStreamingHandler is the middleware handler for bidirectional requests from
+// the server handling perspective.
+func (*Observer) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, stream connect.StreamingHandlerConn) error {
+		rm := getRequestMetadata(ctx)
+		rm.procedure = stream.Spec().Procedure
+		rm.protocol = stream.Peer().Protocol
+
+		wrapped := streamingHandlerConn{
+			base: stream,
+			rm:   rm,
+		}
+		if err := next(ctx, wrapped); err != nil {
+			rm.err = err
+			return err
+		}
+		return nil
+	}
+}
+
+// RequestMetadata represents the metadata gathered during the entire lifecycle
+// of an incoming API request.
+type RequestMetadata struct {
+	startAt          time.Time
+	finishedAt       time.Time
+	duration         time.Duration
+	procedure        string
+	requestURI       string
+	method           string
```

Comment:
method is http method? maybe give it a better name. 
would it make sense to expose http.Request here?